// lumen/lib/gpu/device.js
import * as utl from "../utils.js";
import * as pln from "./pipeline.js";
import * as shd from "./shader.js";
import * as grp from "./group.js";
import * as buf from "./buffer.js";
import * as txr from "./texture.js";
import { Canvas } from "./canvas.js";
import { CommandEncoder } from "./encoder.js";
import { withLabel } from "./utils.js";
export class Device {
    constructor(_wrapped, deviceDescriptor) {
        this._wrapped = _wrapped;
        this.deviceDescriptor = deviceDescriptor;
        this.destructionListeners = [];
        this.recoveryListeners = [];
        this._wrapped.lost.then(info => this.handleDeviceLoss(info));
    }
    async handleDeviceLoss(info) {
        const label = this._wrapped.label ?? 'unlabeled';
        if (info.reason === "destroyed") {
            console.info(`GPU Device '${label}' was lost because it was destroyed.`);
            this.destructionListeners.forEach(listener => listener());
        }
        else {
            console.warn(`GPU Device '${label}' was lost because:`, info.message, ". Attempting to recover ...");
            const { device, descriptor: _ } = await deviceAndDescriptor(this.deviceDescriptor);
            this._wrapped = device;
            console.info(`GPU Device '${label}' was successfully recovered.`);
            this.recoveryListeners.forEach(listener => listener());
        }
    }
    get wrapped() {
        return this._wrapped;
    }
    /**
     * @deprecated Use `wrapped` instead.
     */
    get device() {
        return this.wrapped;
    }
    addDestructionListener(listener) {
        return this.addListener(listener, this.destructionListeners);
    }
    addRecoveryListener(listener) {
        return this.addListener(listener, this.recoveryListeners);
    }
    addListener(listener, listeners) {
        const safeListener = () => {
            try {
                listener();
            }
            catch (e) {
                console.error("Error in device listener: ", e);
            }
        };
        listeners.push(safeListener);
        return () => listeners.splice(listeners.indexOf(safeListener), 1);
    }
    async shaderModules(descriptors, labelPrefix) {
        const result = {};
        for (const k in descriptors) {
            result[k] = await this.shaderModule(withLabel(descriptors[k], labelPrefix, k));
        }
        return result;
    }
    async shaderModule(descriptor) {
        return descriptor.path !== undefined
            ? await this.remoteShaderModule(descriptor.label ?? "shader", descriptor.path, descriptor.templateFunction)
            : await this.inMemoryShaderModule(descriptor.label ?? "shader", descriptor.code, descriptor.templateFunction);
    }
    async loadShaderModule(relativePath, templateFunction = s => s, basePath = "/shaders") {
        return await this.remoteShaderModule(relativePath, relativePath, templateFunction, basePath);
    }
    async remoteShaderModule(label, relativePath, templateFunction = s => s, basePath = "/shaders") {
        const response = await fetch(`${basePath}/${relativePath}`, { method: "get", mode: "no-cors" });
        const rawShaderCode = await response.text();
        return await this.inMemoryShaderModule(label, rawShaderCode, templateFunction);
    }
    async inMemoryShaderModule(label, rawShaderCode, templateFunction = s => s) {
        const shaderCode = templateFunction(rawShaderCode);
        const shaderModule = new shd.ShaderModule(label, this, shaderCode);
        if (await shaderModule.hasCompilationErrors()) {
            throw new Error("Module compilation failed!");
        }
        return shaderModule;
    }
    enqueueCommands(label, encoding) {
        this.enqueue(this.commandBuffer(label, encoding));
    }
    enqueue(...commandBuffers) {
        this.wrapped.queue.submit(commandBuffers);
    }
    commandBuffer(label, encoding) {
        const encoder = new CommandEncoder(label, this);
        encoding(encoder);
        return encoder.finish();
    }
    canvas(element, sampleCount = 1) {
        return new Canvas(this, element, sampleCount);
    }
    texture(descriptor) {
        return new txr.Texture(this, descriptor);
    }
    sampler(descriptor = undefined) {
        return new txr.Sampler(this, descriptor);
    }
    dataBuffers(descriptors, labelPrefix) {
        return buf.DataBuffer.instances(this, descriptors, labelPrefix);
    }
    dataBuffer(descriptor) {
        return buf.DataBuffer.instance(this, descriptor);
    }
    readBuffers(descriptors, labelPrefix) {
        return buf.ReadBuffer.instances(this, descriptors, labelPrefix);
    }
    readBuffer(size, label) {
        return buf.ReadBuffer.instance(this, size, label);
    }
    writeBuffers(descriptors, labelPrefix) {
        return buf.WriteBuffer.instances(this, descriptors, labelPrefix);
    }
    writeBuffer(data, label) {
        return buf.WriteBuffer.instance(this, data, label);
    }
    syncBuffers(descriptors, labelPrefix) {
        return buf.SyncBuffer.instances(this, descriptors, labelPrefix);
    }
    syncBuffer(descriptor) {
        return buf.SyncBuffer.instance(this, descriptor);
    }
    groupLayouts(descriptors, labelPrefix) {
        return grp.BindGroupLayout.instances(this, descriptors, labelPrefix);
    }
    groupLayout(descriptor, label) {
        return grp.BindGroupLayout.instance(this, descriptor, label);
    }
    pipelineLayouts(descriptors, labelPrefix) {
        return pln.PipelineLayout.instances(this, descriptors, labelPrefix);
    }
    pipelineLayout(descriptor, label) {
        return pln.PipelineLayout.instance(this, descriptor, label);
    }
    suggestedGroupSizes() {
        const limits = this.wrapped.limits;
        const wgs = Math.max(limits.maxComputeWorkgroupSizeX, limits.maxComputeWorkgroupSizeY, limits.maxComputeWorkgroupSizeZ);
        const bitCount1D = Math.floor(Math.log2(wgs));
        const bitCount2DY = bitCount1D >>> 1;
        const bitCount2DX = bitCount1D - bitCount2DY;
        const bitCount3DZ = Math.floor(bitCount1D / 3);
        const bitCount3DY = (bitCount1D - bitCount3DZ) >>> 1;
        const bitCount3DX = bitCount1D - bitCount2DY - bitCount3DZ;
        const oneD = 1 << bitCount1D;
        const twoDX = 1 << bitCount2DX;
        const twoDY = 1 << bitCount2DY;
        const threeDX = 1 << bitCount3DX;
        const threeDY = 1 << bitCount3DY;
        const threeDZ = 1 << bitCount3DZ;
        return [
            [oneD],
            [twoDX, twoDY],
            [threeDX, threeDY, threeDZ]
        ];
    }
    async monitorErrors(filter, expression) {
        this.wrapped.pushErrorScope(filter);
        const result = expression();
        const error = await this.wrapped.popErrorScope();
        if (error) {
            throw error;
        }
        return result;
    }
    static async instance(deviceDescriptor = {}) {
        var { device, descriptor } = await deviceAndDescriptor(deviceDescriptor);
        return new Device(device, {
            ...deviceDescriptor,
            gpuDeviceDescriptor: async () => descriptor,
        });
    }
}
async function deviceAndDescriptor(deviceDescriptor) {
    const gpu = utl.required(navigator.gpu, () => "WebGPU is not supported in this environment!");
    const { gpuDeviceDescriptor, xrCompatible } = defaultedDescriptor(deviceDescriptor);
    const adapter = await requestAdapter(gpu, xrCompatible);
    var descriptor = await gpuDeviceDescriptor(adapter);
    const device = await requestDevice(adapter, descriptor);
    return { device, descriptor };
}
function defaultedDescriptor(descriptor) {
    return {
        gpuDeviceDescriptor: descriptor.gpuDeviceDescriptor ?? defaultGPUDeviceDescriptor,
        xrCompatible: descriptor.xrCompatible ?? false
    };
}
async function defaultGPUDeviceDescriptor(adapter) {
    return adapter.info.isFallbackAdapter
        ? utl.failure("The found GPU Adapter is a fallback one that may cause significant responsiveness problems!")
        : {};
}
async function requestAdapter(gpu, xrCompatible) {
    const adapter = utl.required(await utl.timeOut(gpu.requestAdapter({ xrCompatible }), 5000, "GPU Adapter"), () => "No suitable GPU Adapter was found!");
    console.debug("GPU Adapter Info:", adapter.info);
    console.debug("GPU Adapter Features:", [...adapter.features]);
    console.debug("GPU Adapter Limits:", adapter.limits);
    return adapter;
}
async function requestDevice(adapter, deviceDescriptor) {
    const device = utl.required(await utl.timeOut(adapter.requestDevice(deviceDescriptor), 5000, "GPU Device"), () => "Failed to create a GPU Device!");
    console.debug("GPU Device Features:", [...device.features]);
    console.debug("GPU Device Limits:", device.limits);
    return device;
}
