// lumen/lib/gpu/group.js
import { withLabel } from "./utils.js";
export class BindGroupLayout {
    constructor(device, descriptor) {
        this.device = device;
        this.descriptor = descriptor;
        const entryList = [];
        for (const key of Object.keys(descriptor.entries)) {
            const entry = descriptor.entries[key];
            const newEntry = {
                ...entry,
                visibility: entry.visibility.map(s => GPUShaderStage[s]).reduce((a, b) => a | b, 0)
            };
            entryList.push(newEntry);
        }
        this.wrapped = device.wrapped.createBindGroupLayout({
            label: descriptor.label,
            entries: entryList
        });
    }
    instances(descriptors, labelPrefix) {
        const result = {};
        for (const key in descriptors) {
            result[key] = this.instance(withLabel(descriptors[key], labelPrefix, this.descriptor.label, key));
        }
        return result;
    }
    instance(descriptor) {
        return new BindGroup(this, descriptor);
    }
    asEntry(group) {
        return { layout: this, group };
    }
}
export class BindGroup {
    constructor(layout, descriptor) {
        this.layout = layout;
        this.descriptor = descriptor;
        const entryList = [];
        for (const key of Object.keys(descriptor.entries)) {
            entryList.push({
                binding: layout.descriptor.entries[key].binding,
                resource: descriptor.entries[key].asBindingResource()
            });
        }
        this.wrapped = layout.device.wrapped.createBindGroup({
            label: descriptor.label,
            layout: layout.wrapped,
            entries: entryList
        });
    }
}
