// lumen/lib/gpu/group.js
export class BindGroupLayout {
    constructor(device, label, descriptor) {
        this.device = device;
        this.label = label;
        this.descriptor = descriptor;
        const entryList = [];
        for (const key of Object.keys(descriptor.entries)) {
            const entry = descriptor.entries[key];
            const newEntry = {
                ...entry,
                visibility: entry.visibility.map(s => GPUShaderStage[s]).reduce((a, b) => a | b, 0)
            };
            entryList.push(newEntry);
        }
        this.wrapped = device.wrapped.createBindGroupLayout({
            label,
            entries: entryList
        });
    }
    instance(label, descriptor) {
        return new BindGroup(label, this, descriptor);
    }
    instances(descriptors) {
        const result = {};
        for (const key in descriptors) {
            result[key] = this.instance(key, descriptors[key]);
        }
        return result;
    }
    asEntry(group) {
        return { layout: this, group };
    }
}
export class BindGroup {
    constructor(label, layout, descriptor) {
        this.label = label;
        this.layout = layout;
        this.descriptor = descriptor;
        const entryList = [];
        for (const key of Object.keys(descriptor.entries)) {
            entryList.push({
                binding: layout.descriptor.entries[key].binding,
                resource: descriptor.entries[key].asBindingResource()
            });
        }
        this.wrapped = layout.device.wrapped.createBindGroup({
            label: `${layout.label}@${label}`,
            layout: layout.wrapped,
            entries: entryList
        });
    }
}
