// gear/lib/files.js
export async function fetchTextFile(url) {
    return fetch(url, { method: "get", mode: "no-cors" }).then(response => response.text());
}
export async function fetchTextFiles(files, path = ".") {
    const result = {};
    const keys = Object.keys(files);
    const promises = keys.map(k => requestTextFile(k, `${path}/${files[k]}`));
    for (let [key, promise] of promises) {
        result[key] = await promise;
    }
    return result;
}
function requestTextFile(key, url) {
    return [key, fetch(url, { method: "get", mode: "no-cors" }).then(response => response.text())];
}
export function save(url, contentType, fileName) {
    const anchor = document.createElement("a");
    anchor.href = url;
    anchor.type = contentType;
    anchor.target = '_blank';
    anchor.download = fileName;
    anchor.click();
}
export class FileSelector {
    constructor(multiple, mimeTypes) {
        this.multiple = multiple;
        this.mimeTypes = mimeTypes;
        this.input = document.createElement("input");
        this.input.type = "file";
        this.input.multiple = multiple;
        this.input.accept = mimeTypes.join(",");
    }
    static create() {
        return new FileSelector(false, []);
    }
    allowMultipleFiles() {
        return this.multiple ? this : new FileSelector(true, this.mimeTypes);
    }
    disallowMultipleFiles() {
        return this.multiple ? new FileSelector(false, this.mimeTypes) : this;
    }
    ofType(...mimeTypes) {
        return new FileSelector(this.multiple, mimeTypes);
    }
    async select() {
        return new Promise(resolve => {
            this.input.onchange = () => resolve(this.files());
            this.input.click();
        });
    }
    files() {
        return this.input.files && this.input.files.length > 0
            ? (this.multiple === true ? [...this.input.files] : [this.input.files[0]])
            : [];
    }
}
