// gear/lib/loops/loop.js
import { Button, VirtualKey } from './button.js';
import { Keyboard } from './keyboard.js';
import { Pointer } from './pointer.js';
import { CanvasRecorder, CanvasSizeManager } from './canvas.js';
import { FrequencyMeter } from './perf.js';
import { required } from '../utils.js';
export function newLoop(loopLogic, inputDescriptor) {
    return new LoopImpl(loopLogic, inputDescriptor);
}
class LoopImpl {
    constructor(loopLogic, loopDescriptor) {
        this.loopLogic = loopLogic;
        this.loopDescriptor = loopDescriptor;
        this.keyboard = new Keyboard();
        this.canvasSizeManager = new CanvasSizeManager(true);
        this.render = () => { };
        this.frequencyMeter = this.loopDescriptor.output.fps
            ? FrequencyMeter.create(this.loopDescriptor.output.fps.periodInMilliseconds ?? 1000, this.loopDescriptor.output.fps.element)
            : new FrequencyMeter(1000, () => { });
        this._paused = false;
        const deferredInputOps = [];
        const deferredOutputOps = [];
        this.keys = this.createKeys(loopDescriptor, deferredInputOps);
        this.pointers = this.createPointers(loopDescriptor, deferredInputOps);
        this.canvases = this.getCanvases(loopDescriptor, deferredOutputOps);
        const inputWiring = loopLogic.inputWiring(this, this, this);
        const outputWiring = loopLogic.outputWiring(this);
        this.render = outputWiring.onRender;
        deferredInputOps.forEach(op => op(inputWiring));
        deferredOutputOps.forEach(op => op(outputWiring));
    }
    getCanvases(loopDescriptor, deferredOps) {
        const canvases = {};
        if (loopDescriptor.output.canvases) {
            for (const canvasName of keysOf(loopDescriptor.output.canvases)) {
                const canvasDescriptor = loopDescriptor.output.canvases[canvasName];
                const canvas = this.getCanvasElement(canvasDescriptor);
                canvases[canvasName] = {
                    element: canvas,
                    recorder: new CanvasRecorder(canvas)
                };
                deferredOps.push(loopWiring => this.wireCanvas(canvasName, canvas, loopWiring));
            }
        }
        return canvases;
    }
    getCanvasElement(canvasDescriptor) {
        return required(document.getElementById(canvasDescriptor.element));
    }
    wireCanvas(canvasName, canvas, loopWiring) {
        if (!loopWiring.canvases) {
            return;
        }
        const canvasWiring = loopWiring.canvases[canvasName];
        if (canvasWiring && canvasWiring.onResize) {
            this.canvasSizeManager.observe(canvas, canvasWiring.onResize);
        }
    }
    createKeys(loopDescriptor, deferredOps) {
        const keys = {};
        if (loopDescriptor.input.keys) {
            for (const keyName of keysOf(loopDescriptor.input.keys)) {
                const keyDescriptor = loopDescriptor.input.keys[keyName];
                const button = this.newButton(keyDescriptor, loopDescriptor.output.styling, loopDescriptor.input.keys);
                keys[keyName] = button;
                deferredOps.push(loopWiring => {
                    this.wireButton(keyName, button, loopWiring);
                });
            }
        }
        return keys;
    }
    wireButton(keyDescriptorKey, button, loopWiring) {
        const wiring = loopWiring.keys ? loopWiring.keys[keyDescriptorKey] : null;
        if (!wiring) {
            return;
        }
        if (wiring.onPressed) {
            const onPressed = wiring.onPressed;
            button.register(b => {
                if (b.pressed) {
                    onPressed();
                }
            });
        }
        if (wiring.onReleased) {
            const onReleased = wiring.onReleased;
            button.register(b => {
                if (!b.pressed) {
                    onReleased();
                }
            });
        }
        if (wiring.onChange) {
            const onChange = wiring.onChange;
            button.register(b => {
                onChange();
            });
        }
    }
    createPointers(loopDescriptor, deferredOps) {
        const pointers = {};
        if (loopDescriptor.input.pointers) {
            for (const pointerDescriptorKey of keysOf(loopDescriptor.input.pointers)) {
                const pointerDescriptor = loopDescriptor.input.pointers[pointerDescriptorKey];
                const pointer = new Pointer(pointerDescriptor.element);
                pointers[pointerDescriptorKey] = pointer;
                deferredOps.push(loopWiring => {
                    this.wirePointer(pointerDescriptorKey, pointer, loopWiring);
                });
            }
        }
        return pointers;
    }
    wirePointer(pointerDescriptorKey, pointer, loopWiring) {
        const wiring = loopWiring.pointers[pointerDescriptorKey];
        if (!wiring) {
            return;
        }
        if (wiring.defaultDraggingTarget) {
            pointer.draggingTarget = wiring.defaultDraggingTarget;
        }
        if (wiring.onMoved) {
            const onMoved = wiring.onMoved;
            pointer.register(p => onMoved());
        }
        if (wiring.primaryButton) {
            const buttonWiring = wiring.primaryButton;
            if (buttonWiring.onPressed) {
                const onPressed = buttonWiring.onPressed;
                pointer.primary.register(b => {
                    if (b.pressed) {
                        onPressed();
                    }
                });
            }
            if (buttonWiring.onReleased) {
                const onReleased = buttonWiring.onReleased;
                pointer.primary.register(b => {
                    if (!b.pressed) {
                        onReleased();
                    }
                });
            }
            if (buttonWiring.onChange) {
                const onChange = buttonWiring.onChange;
                pointer.primary.register(() => {
                    onChange();
                });
            }
        }
    }
    get animationPaused() {
        return this._paused;
    }
    set animationPaused(p) {
        this._paused = p;
    }
    run() {
        if (LoopImpl.activeLoop !== this) {
            LoopImpl.activeLoop = this;
            for (const key of keysOf(this.pointers)) {
                const pointer = this.pointers[key];
                pointer.use();
            }
            this.keyboard.use();
            this.nextFrame();
        }
    }
    nextFrame() {
        requestAnimationFrame(time => {
            const delta = this.frequencyMeter.tick(time);
            if (!this.animationPaused) {
                this.loopLogic.animate(time, delta, this);
            }
            this.render();
            if (this === LoopImpl.activeLoop) {
                this.nextFrame();
            }
        });
    }
    newButton(keyDescriptor, styling, keyDescriptors) {
        const pressedClass = styling?.pressedButton ?? "";
        const releasedClass = styling?.releasedButton ?? "";
        const virtualButtons = keyDescriptor.virtualKeys !== undefined
            ? [...document.querySelectorAll(keyDescriptor.virtualKeys)].filter(e => e instanceof HTMLElement)
            : [];
        const allShortcuts = [];
        Object.keys(keyDescriptors).forEach(k => allShortcuts.push(...keyDescriptors[k].physicalKeys));
        const button = Button.anyOf(...keyDescriptor.physicalKeys.map(shortcut => {
            const complements = allShortcuts
                .map(otherShortcut => {
                const complement = otherShortcut.filter(k => shortcut.indexOf(k) < 0);
                return (complement.length + shortcut.length) == otherShortcut.length ? complement : [];
            })
                .filter(set => set.length > 0);
            return Button.allOf(...shortcut.map(k => this.keyboard.key(k)), ...complements.map(c => Button.allOf(...c.map(k => this.keyboard.key(k))).not()));
        }), ...virtualButtons.map(e => new VirtualKey(e)));
        if (virtualButtons.length > 0) {
            button.register(b => {
                if (b.pressed) {
                    virtualButtons.forEach(b => {
                        b.className = `${b.className.replace(releasedClass, "")} ${pressedClass}`.trim();
                    });
                }
                else {
                    virtualButtons.forEach(b => {
                        b.className = `${b.className.replace(pressedClass, "")} ${releasedClass}`.trim();
                    });
                }
            });
        }
        return button;
    }
}
LoopImpl.activeLoop = null;
function keysOf(object) {
    return Object.keys(object);
}
